### PR TITLE
Change unused_param check default to Error

### LINKED ISSUE #20567: Change default value of unused_parameter flag from Warning to Error
## Reason

1. It is more reasonable to change the default value of unused_parameter flag from warning to error. 
2. In multiapp, the sub app will not report an error for unused_parameter even if the main app sets the option to Error.
## Design

1. Change the default value.
2. Let sub app  use the same flag as the main app uses
## Impact

The default should only affect applications or modules that don’t have an explicit setting already.

### PR DESCRIPTION
**Pull request by jiangwen84**
closes #20567

### CODE REVIEW

**Comment by GiudGiud** [comment]
@idaholab/moose-ccb opinions on this? This does have pretty big consequences.

I very much like being able to change input parameters, to try out things, and not worry about what becomes unused
I also end up debugging people's inputs where they had a mispelled parameter and it only returned a warning, and the wrong result
So overall I'm in favor of this, even though it could be a possible burden on users.

**Comment by lindsayad** [comment]
I think @aprilnovak had some concerns about this

**Comment by gardnerru** [comment]
I expected this to be contentious. Here is more background.

This was found when I was setting up a very large Monte Carlo with BISON. I made a BISON input, with the proper parameters, and tested it. I then made the driver file and tested it with a small sample size. This ran and the SamplerData CSV output showed that the params were changing as expected. With this, and that I know that BISON errors on unused, I felt my diligence was done. Turns out that there was a typo in the MC driver file that was trying to pass an unusable parameter to the BISON input, and MOOSE overrode BISON and tossed a warning. I know that someone here is going to say that I should have read my logs better. That might be so, but I wasnt expecting a behavior change and I would still be here when I saw it.  

Multiapp might be a special case but at the very least here, the MOOSE main app should use whatever the sub's defaults are. This is a trap.

Does anyone have an _actual_ reason for having an unused parameter in their input file? I get it for the old days but when you are moving more in to production, allowing unused seems like a bad idea. End of soap box.

**Comment by friedmud** [comment]
Firstly, a question: @gardnerru , what app were you actually running as your main app (as in, what app did you run on the command-line?).

Secondly: I think I agree here (although I feel like I was on the other side of this when it was first decided).  The issue surrounds swapping out objects.  For instance, the current behavior is useful if you want to swap from a `FluxBC` to a `NeumannBC` and then have the ability to swap back (without commenting a bunch of stuff out and then uncommenting).

However, this just isn't the norm to do anymore.  People are changing their objects much less often these days.  The only case I can think of where people would do this often is in the Material system (swapping in and out simpler solid-mechanics materials for instance).... and if BISON has already had this as an error for a long time and no one has complained then it must be "fine".

I also think that the current behavior was useful for Peacock... which just left old parameters hanging around when you would swap objects - so that you could swap them back.  But no one uses that anyway......

I'm for changing this to be an error (until someone has a good argument in the other direction!).

**Comment by gardnerru** [comment]
./bison-opt -i monte_carlo.i , where monte_carlo.i is a Stochastic module input calling a BISON input.

Hehe, using BISON as the gold standard? Careful, we might get a big head.

**Comment by aprilnovak** [comment]
Thanks @lindsayad, yes this would negatively impact my workflow. I use very complex, multi-application models that share some common input. Say, Apps 1 and 2 need a parameter `diameter`, whereas Apps 2 and 3 need a parameter `conductivity`. Rather than try to remember precisely where each parameter is used in my inputs (in case I want to change them), I have a `common_input.i` that I pass via the command line to the various input files that need it, like this:

```
cardinal-opt -i common_input model.i
```

This will give unused parameter warnings, since some application inputs only need a subset of the parameters. I think from Slack I learned that I can always set a different warn/error setting in Cardinal, but I'm not sure if I could then use my `common_input.i` if I'm, say, dynamically linking to a BISON input which does set the error.

I'd be okay with this change if the preamble constants at the top of a file could be ignored.

**Comment by tophmatthews** [comment]
+1 to @aprilnovak , I do the same thing (though not so complicated) and it's a pain for these out-of-block lines (or whatever they are called when they're not in []). However, the error in BISON for input parameters not being necessary has saved me in the past, so I think @aprilnovak has the right balance.

**Comment by lindsayad** [comment]
I think an input file option controlling unused parameter warn/error would be sublime. Same with deprecation I think

**Comment by gardnerru** [comment]
So set the moose default to error on unused params and then let the user command line flag if they need otherwise? Im fine with that. I think it may leave a small hole but it wont catch me. :)

**Comment by andrsd** [comment]
Can we do an environment variable like `MOOSE_UNUSED_PARAMS={NONE|WARNING|ERROR}` and get the desired behavior via that?

**Comment by friedmud** [comment]
@andrsd I don't see why we couldn't do that _as well_... no reason not to allow all the options.

**Comment by joshuahansel** [comment]
> MOOSE overrode BISON and tossed a warning

Is there any reason we can't fix this part directly instead of changing defaults?

**Comment by joshuahansel** [comment]
> I don't see why we couldn't do that as well... no reason not to allow all the options.

@friedmud I'd find it a bit confusing to have all of those options because it wouldn't be obvious which methods take precedence.

**Comment by GiudGiud** [comment]
Precedence is:
Command line > input file > environment variable > app default
Any other order is wrong :)

**Comment by friedmud** [comment]
@GiudGiud is correct.  That is how all such options should work.

**Comment by friedmud** [comment]
More options are not bad - as long as the precedence is documented.

**Comment by bwspenc** [comment]
I'll just chime in here in support of making error the default.

Regarding having an environment variable, I'm not aware of anything else in application execution that is controlled by environment variables. Do we actually have environment variables that control things like this now? I'd be opposed to having this be the only one, but if we have others already, then sure, do it.

**Comment by bwspenc** [comment]
The downside to environment variables is that if we start using them widely, we're going to start having the program behave differently for different people. The switch we joke about to give different behavior for @tophmatthews will become a reality. I'm not saying there's never a place for it, but it will come back to bite us when somebody's complaining that a problem fails for them but it works just fine for us when we try running it. I can also see a developer with unused variables set to warnings in their environment variable checking in a test with an unused variable that fails when it runs on our build machines. Not the end of the world, but it could waste someone's time trying to figure out what's going on.

**Comment by permcody** [comment]
We just need to create another environment variable in secret to overrule
all the others! I’m sure no abuse of such a system could occur as long as
we possess it.

**Comment by lindsayad** [comment]
we could create a `IGNORE_ENVIRONMENT_VARIABLES` option for users, and then for system administrators who get tired of debugging cases where a user has set that, we could create for them a `IGNORE_IGNORE_ENVIRONMENT_VARIABLES`

**Comment by tophmatthews** [comment]
> The switch we joke about to give different behavior for @tophmatthews will become a reality

I _KNEW_ IT!

**Comment by gardnerru** [comment]
So what is the move here? Stock-holders vote or Thunderdome rules?

**Review by lindsayad — CHANGES_REQUESTED** [review]
While the default change is good, we must support being able to make unused parameters *not* errors

**Comment by gardnerru** [comment]
@permcody was there a Tuesday morning chat about this?

**Comment by permcody** [comment]
Nope… my fault. Many of us are back in the building. I’ll bring it up
tomorrow!

**Comment by YaqiWang** [comment]
I am late for this conversation but I am very supportive on what @bwspenc said on using environmental variables.

**Comment by GiudGiud** [comment]
He forgot again. I ll remind him this afternoon haha

File: test/tests/misc/jacobian/simple.i [hunk 1]
```diff
@@ -37,7 +37,6 @@
   [./diffu]
     type = WrongJacobianDiffusion
     variable = u
-    error = factor
```
Comment (lindsayad) [inline test/tests/misc/jacobian/simple.i:40]: what was the reason for this change?

**Comment by lindsayad** [comment]
we're changing the default to error, but we need to allow downgrading. @socratesgorilla I think is going to try to push that change soon

**Comment by permcody** [comment]
@socratesgorilla - This is the PR that I was talking about in staff meeting. Need to get it over the finish line.

**Comment by GiudGiud** [comment]
Can you please double check that `allow_unused` in tests configuration works? Having issues with this.
maybe it s not a thing

**Comment by gardnerru** [comment]
Asleep just before the finish line? I feel that I read a story like this a long time ago

**Comment by permcody** [comment]
@socratesgorilla?

**Comment by socratesgorilla** [comment]
Was working on this and then got sidetracked by moving stuff - taking another look now

**Comment by GiudGiud** [comment]
@jiangwen84 can we close this?
https://github.com/idaholab/moose/pull/21821 was merged

**Comment by GiudGiud** [comment]
Capability was merged in another PR

**Comment by lindsayad** [comment]
> Yes. Close it

image

**Comment by lindsayad** [comment]
I always like an option that you can check into version control such that
an input when cloned anywhere behaves more or less the same anywhere, e.g.
the input file option. But I would support having an environment variable
*in addition* ... which I think an input (or cli argument) would override

**Comment by GiudGiud** [comment]
so we want to be able to control this with:

- [ ] input file
- [ ] CLI
- [ ] environment variable

I think it's time many of our CLIs make it to the input file anyway?

**Comment by lindsayad** [comment]
I thought we wanted to forego the environment variable option?

**Comment by friedmud** [comment]
I wanted all the options - but I'll settle for just command-line for now.

**Comment by socratesgorilla** [comment]
I don't know why this conversation is happening on a closed PR but I believe it works on CLI and input file for sure.

**Pull request closed without merging**